### PR TITLE
clickable color swatch with QColorDialog

### PR DESCRIPTION
**Pull request by tlambert03**
# Description
This PR turns the color swatch (in, e.g., the points layer controls) into a clickable widget that raises a color picker dialog, (as proposed by @sofroniewn [here](https://github.com/napari/napari/issues/762#issuecomment-563048345)).  I would propose that we remove the QComboBox color picker altogether and either have just the clickable swatch, or a LineEdit that can convert arbitrary input into our internal color representation as in #782.  This PR is just a stub/example for now, and will wait for #782 to merge before final integration (if the general GUI idea is approved here).

ezgif-2-107ce933206f

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [x] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Comment by sofroniewn** [comment]
Dropping the `QComboBox` sounds good. I'm open to either just having the swatch like Photoshop / Illustrator does, or the `QLineEdit` which we can type into / will show the string representation after we change colors.

Waiting till after #782 makes sense too

**Comment by kevinyamauchi** [comment]
This is cool! I am becoming more and more convinced that @tlambert03 is a wizard.

I agree that we should drop the `QComboBox`. Outside of the standard red, green, blue, etc., I don't think the string representations (e.g., hex codes or `chartruesse` 😛 ) are super informative and I think most people will prefer to pick visually.

**Comment by sofroniewn** [comment]
Ok yeah - maybe we just have the swatch, especially if the string name is visible somewhere inside the color dialog (even if not though, probably still ok not to have the QLineEdit - keeps things simpler and cleaner) but I defer to the preference of @tlambert03.

And yes, another amazing addition!!!

**Comment by jni** [comment]
(a) this is super brilliant
(b) I still like the ability to pick colours by name somehow, which might be useful where exact reproducibility is needed. But perhaps it's fine to leave such functionality to the console.
(c) this gave me the idea that we could possibly make the colormap picker equally visual, with a grid of gradients. 🤯

**Comment by sofroniewn** [comment]
It doesn't look like this PR has conflicts with #686 which had the conversions like `edge_color `-> `current_edge_color` and `edge_colors` -> `edge_color` but just wanted to tag here in case

**Comment by tlambert03** [comment]
Thanks to @HagaiHargil's work in #782, polishing up this PR was made much easier.  I updated it with a generalized `QColorSwatchEdit` widget, which combines a colorswatch that raises an interactive`QColorDialog` on click with a `QLineEdit` that will both tab-autocomplete "standard" color names (any key from `vispy.color.get_color_dict()`), as well as accept hex values or even rgb(a) values with or without parenthesis (accomplishing goal (b) from @jni above).

you can try it out by fetching this PR and using `examples/add_points.py`:

Untitled


**2 questions:**
1. do we prefer Qt Color dialog over an OS-native color picker?  (has the benefit of being consistent across platforms, and matching the rest of napari look and feel, but the downside of perhaps breaking user expectations)
2. I updated the points layer controls to use this.  Is there anywhere else we have color pickers?

**Comment by sofroniewn** [comment]
> I updated the points layer controls to use this. Is there anywhere else we have color pickers?

We use them in shapes and vectors, but we still need to get the #898 refactor from @HagaiHargil in before it might be as easy to adapt that here. You can take a look @tlambert03

**Comment by tlambert03** [comment]
> we still need to get the #898 refactor from @HagaiHargil in before it might be as easy to adapt that here

ah right.  Well there's no particular reason to wait on that for this PR (this one stands alone).  In fact, it might be a good idea to get this in so they can follow this pattern there if desired.

**Comment by sofroniewn** [comment]
> ah right. Well there's no particular reason to wait on that for this PR (this one stands alone). In fact, it might be a good idea to get this in so they can follow this pattern there if desired.

Agreed. I'm playing with this now. I'm finding some unexpected behaviour if i enter something bad which leads to an "unknown" warning and then start clicking to add points it defaults to "transparent" which I don't think it should do. I wonder if instead of giving "unknown"  it should just go back to what it was before?

Also is there a way to make it easy to select transparent from the dialog? Like a white square with a red diagonal line through it? Transparent also doesn't seem to have an autocomplete like the other colors do.

I do have some concerns about the stylings of the QDialog, there are a lot of options so it might seem overwhelming. I'm ok with the non-native feel. In Illustrator it is non-native. I think we can always revisit the stylings if we get a full-time designer working on the project.

I'm going to keep looking this over, but I think it is close

File: napari/_qt/qt_color_dialog.py
Comment (sofroniewn) [inline napari/_qt/qt_color_dialog.py]: Ah looking at the code I see why what I mentioned in the main thread happened, but I did find it unexpected as a user

Comment (tlambert03) [reply napari/_qt/qt_color_dialog.py]: no problem.  we can either make it black, or go back to what it was before.

File: napari/_qt/qt_color_dialog.py
Comment (sofroniewn) [inline napari/_qt/qt_color_dialog.py:109]: Just so I understand if this where the autocomplete functionality is happening? I've not used autocomplete in Qt before, could you just say a little something about where / how that is working (and say why I don't seem to get transparent to autocomplete)

Comment (sofroniewn) [reply napari/_qt/qt_color_dialog.py:109]: ah I see it below - sorry I should be doing this in review mode!!

File: napari/_qt/qt_color_dialog.py
Comment (sofroniewn) [inline napari/_qt/qt_color_dialog.py]: Ah here is the autocomplete, which I think needs transparent in there

File: napari/_qt/qt_color_dialog.py
Comment (sofroniewn) [inline napari/_qt/qt_color_dialog.py]: I like QtColorDialog, gives us more flexibility in the long run and is how illustrator looks. As noted in the main thread, we could get help from a designer in the future too

File: napari/_qt/qt_color_dialog.py
Comment (sofroniewn) [reply napari/_qt/qt_color_dialog.py]: i'd probably have it go back to what it was before, seems like the least confusing. might be worth trying what does it look like to quickly flash an "unknown" before reseting it, but that might be weird too. I'm not so so worried about this as we can improve it in the future

Comment (tlambert03) [reply napari/_qt/qt_color_dialog.py]: easily done.. changing now

**Review by sofroniewn — APPROVED** [review]
This looks good to me, I am approving though I'd like the minor changes I mention above made if possible. Then maybe @kevinyamauchi and @jni want to have a look.

I love that this PR didn't actually effect the points layer itself. That's good modularity!!

**Comment by sofroniewn** [comment]
Also @tlambert03 I see the 3.7 test is failing, which I think is independent of this PR - and I think has been discussed elsewhere, but can you link me to that discussion / give me an update on plans to fix? It looks like it is in in a `test_view` is it possible because we aren't using an np.random.seed and so sometimes are generating weird data?

**Comment by tlambert03** [comment]
> Also @tlambert03 I see the 3.7 test is failing, which I think is independent of this PR - and I think has been discussed elsewhere, but can you link me to that discussion / give me an update on plans to fix? It looks like it is in in a `test_view` is it possible because we aren't using an np.random.seed and so sometimes are generating weird data?

I really don't know, and I'm not sure I can be the one working on that one one. I did try to figure it out at one point (i think it's a similar phenomenon to #925) but it's beyond me at the moment.  I think @ttung was starting to look into that bug too... maybe he has time to pick it up?

**Comment by tlambert03** [comment]
ok. transparent is added as an option to the autocomplete choices, and entering an invalid string sends you back to the previous setting.

>Also is there a way to make it easy to select transparent from the dialog? Like a white square with a red diagonal line through it?

the built in `QColorDiag` is actually not that flexible... so I'm not sure how we would add this sort of thing (or generally adjust the look and feel even if we had a design in mind).  Note however, there is a `Alpha Channel:` option in the dialog that you can set to 0.

**Comment by sofroniewn** [comment]
> I really don't know, and I'm not sure I can be the one working on that one one. I did try to figure it out at one point (i think it's a similar phenomenon to #925) but it's beyond me at the moment. I think @ttung was starting to look into that bug too... maybe he has time to pick it up?

Yup thats fine @tlambert03 your focus should be elsewhere, thanks for the link.

> the built in QColorDiag is actually not that flexible... so I'm not sure how we would add this sort of thing (or generally adjust the look and feel even if we had a design in mind). Note however, there is a Alpha Channel: option in the dialog that you can set to 0.

Ok that's fine for me.

I have pulled this and it all works great. I think this is good to go then, but as noted above I will let @kevinyamauchi and @jni have a look before merge

**Comment by tlambert03** [comment]
> Yup thats fine @tlambert03 your focus should be elsewhere, thanks for the link.

actually... now you got me thinking about it :) I think one potential fix is to use @ttung's `viewermodel_factory()` more.  The `view_layer_type` test utility function might need to be adjusted.  taking a look now

**Comment by tlambert03** [comment]
it's hard to use his factory while testing the `view_*` methods, but i think the key in tony's method was calling `view.shutdown()` (rather than `viewer.window.close()`). 
I'm trying stuff here and will PR if it works: https://github.com/tlambert03/napari/pull/7

File: napari/_qt/layers/qt_points_layer.py
Comment (jni) [inline napari/_qt/layers/qt_points_layer.py:167]: Wow, yeah, this is so great now!

File: napari/_qt/qt_color_dialog.py
Comment (jni) [inline napari/_qt/qt_color_dialog.py:143]: I think the convention, for better or worse, is to have the `__init__` parameters in this docstring, as well as the public attributes.

File: napari/_qt/qt_color_dialog.py
Comment (jni) [inline napari/_qt/qt_color_dialog.py:166]: Should this be implemented as a property?

File: napari/_qt/qt_color_dialog.py
Comment (jni) [inline napari/_qt/qt_color_dialog.py:264]: Could you add Parameters and Attributes to this docstring? e.g. I don't know what `initial` does looking at the `__init__` signature.

Comment (jni) [reply napari/_qt/qt_color_dialog.py:264]: (It might benefit from a rename to "initial_color"

File: napari/_qt/qt_color_dialog.py
Comment (jni) [inline napari/_qt/qt_color_dialog.py]: Very nice. =)

Can you add a Parameters section here, specifying the class of `event`? ie if I want to modify the behaviour of this function, how do I see what I can and can't do with `event`?

File: napari/layers/utils/color_transformations.py
Comment (jni) [inline napari/layers/utils/color_transformations.py]: Hmmm, I don't like this change... It sneaks Qt into our models code. It seems like the wrong solution. If you need a signature that takes in a QColor in addition to all of the above, I suggest defining a new type `QColorType = U[ColorType, QColor]` in our Qt files... If you want `standardize_color` to take in QColors, then I suggest modifying the dictionary from the view files.

**Review by jni — CHANGES_REQUESTED** [review]
@tlambert03 I've made a few minor comments, but request a major change, which is to remove all Qt imports from our non-Qt color-handling code. I think it should be feasible to keep all this logic on the Qt side of things.

File: napari/layers/utils/color_transformations.py
Comment (tlambert03) [reply napari/layers/utils/color_transformations.py]: don't need it to.  I like the idea of pulling out all the qt stuff... just wasn't thinking.  Will update.

File: napari/_qt/qt_color_dialog.py
Comment (tlambert03) [reply napari/_qt/qt_color_dialog.py:166]: I'm always torn when doing Qt stuff, whether it's better to mimic the Qt API or the pythonic API.

Comment (jni) [reply napari/_qt/qt_color_dialog.py:166]: I'm inclined to go with Pythonic in ambiguous cases. =) Also, I think in cases where we are creating a class that we will *mostly* interact with using our own methods, rather than the built-in Qt methods, we should use a Pythonic API.

File: napari/_qt/qt_color_dialog.py
Comment (tlambert03) [reply napari/_qt/qt_color_dialog.py:143]: really? I thought it was the other way around...  [pep 257](https://www.python.org/dev/peps/pep-0257/#multi-line-docstrings) seems to suggest that public methods and instance vars can go in the class docstring, but the constructor parameters go in the `__init__` docstring.  in any case, i can add the public stuff to the class docstring.

Comment (jni) [reply napari/_qt/qt_color_dialog.py:143]: That's disappointing that it goes against PEP 257, but it's what I've seen everywhere else, e.g. sklearn:

https://github.com/scikit-learn/scikit-learn/blob/244d118ef77a513b487b95b721179c395cbc1660/sklearn/ensemble/_forest.py#L850

Comment (jni) [reply napari/_qt/qt_color_dialog.py:143]: (And it makes sense for sphinx rendering...)

Comment (tlambert03) [reply napari/_qt/qt_color_dialog.py:143]: yeah, that's a bummer... I don't love it actually.  but I guess the hoops that sphinx makes you jump through with `__init__` justifies it.

**Comment by tlambert03** [comment]
thanks for the reviews... all set for another look

File: napari/_qt/qt_color_dialog.py
Comment (jni) [inline napari/_qt/qt_color_dialog.py:182]: `@color.setter`?

**Review by jni — APPROVED** [review]
@tlambert03 just one more minor comment, but approving nonetheless.

File: napari/_qt/qt_color_dialog.py
Comment (tlambert03) [reply napari/_qt/qt_color_dialog.py:182]: please no, much hard to use to connect to signals.  i know we don't have to do everything the Qt way, but I think getting too far away from the standard qt patterns and conventions starts to just confuse anyone who _is_ familiar with it.

Comment (jni) [reply napari/_qt/qt_color_dialog.py:182]: Oh ok! Sorry, "anyone who is familiar with it" and myself is an empty set. =P The question mark was genuine! =)

**Comment by sofroniewn** [comment]
If this is done and @tlambert03 you want to merge then go ahead and I'll update the release notes #994 accordingly and this can go in `0.2.12`. I don't plan to merge #994 until tomorrow anyway to let some more of the team look it over so you have time

File: napari/_qt/qt_color_dialog.py
Comment (tlambert03) [reply napari/_qt/qt_color_dialog.py:182]: lol.  ok :)  yeah... just cause it's funny, here's what using a `@setter` would do to line 179 just above it:
```python
popup.colorSelected.connect(self.setColor)

popup.colorSelected.connect(lambda color: self.__class__.color.fset(self, color))
```

**Pull request merged**
